patches/server/0977-Fix-multiple-spectator-bugs.patch

```diff
@@ -0,0 +1,61 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Fritz Windisch <friwidev@gmail.com>
+Date: Thu, 15 Jun 2023 17:10:22 +0200
+Subject: [PATCH] Fix multiple spectator bugs
+
+Fixes MC-261799: Spectator mode not teleporting viewers to targets in other worlds or viewers bug around (stuck) when a target changes world
+Fixes MC-107113: Spectator mode viewers stuck when target is teleported more than viewing distance
+Fixes MC-148993: Chunks go invisible when target moves too far over time while viewing (render POI on client is not updated)
+
+diff --git a/src/main/java/net/minecraft/server/level/ChunkMap.java b/src/main/java/net/minecraft/server/level/ChunkMap.java
+index 9209b598d7168b82574e4800056b8b9f84265dd0..2b37ccc8e2619677179b17a6c5bd758bd5f95150 100644
+--- a/src/main/java/net/minecraft/server/level/ChunkMap.java
++++ b/src/main/java/net/minecraft/server/level/ChunkMap.java
+@@ -1535,6 +1535,7 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
+                 double d1 = vec3d_dx * vec3d_dx + vec3d_dz * vec3d_dz; // Paper
+                 double d2 = d0 * d0;
+                 boolean flag = d1 <= d2 && this.entity.broadcastToPlayer(player);
++                flag |= player.getCamera() == this.entity; // Paper - Make entities always visible for spectators, even when teleported far away (MC-107113)
+ 
+                 // CraftBukkit start - respect vanish API
+                 if (!player.getBukkitEntity().canSee(this.entity.getBukkitEntity())) {
+diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
+index 9d46536f80b5b3e6641fd377c02166a431edfd77..a6b55a8a527816230f2d1853c023161c992ca1da 100644
+--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
++++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
+@@ -11,6 +11,7 @@ import com.mojang.serialization.Dynamic;
+ import java.net.InetSocketAddress;
+ import java.net.SocketAddress;
+ import java.util.Collection;
++import java.util.Collections;
+ import java.util.Iterator;
+ import java.util.List;
+ import java.util.Objects;
+@@ -732,8 +733,25 @@ public class ServerPlayer extends Player {
+ 
+         if (entity != this) {
+             if (entity.isAlive()) {
+-                this.absMoveTo(entity.getX(), entity.getY(), entity.getZ(), entity.getYRot(), entity.getXRot());
+-                this.serverLevel().getChunkSource().move(this);
++                // Paper start - Fix spectator on cross-world teleports (MC-261799)
```

Comment:
This is not the right fix and doesn't work for me. The reason it doesn't work is that mojang explicitly prevent spectating across dimension and you can easily see it in the following line: ServerPlayer#teleportTo L2225~ (line doesn't match since i have modified the source). The way you would fix that (i think) would be either:
* by changing the ServerPlayer#setCamera saving the camera before the world tp and then resetting the camera back to what we expect here but then the spectator events doesn't really work.

* or by simply ignoring this restriction only for this specific call (all the teleportTo called with the bukkit spectate reason) normally this doesn't collide with the fact that plugin can change the teleport reason since they doesn't seems to use this method). I recommend this one.

You can then update the world where chunkSource#move is called.

Reply:
You obviously can not start spectating while your target is in another world. But I see no issue with this to be honest and consider it more like a thing that should be implemented in the /spectate command if you wish to have this functionality. Then also all your events will continue working.

Reply:
Hum, i think you shouldn't state that this fix MC-261799 since it does not and you can't fix this only for the spectate command since plugin could call this also.

Just to be clear your fix here is just when the entity that the player spectate changes of world ? and not the reverse way (MC-261799) ?

Reply:
It appears to be so. I guess it would be possible to fix the spectate issue though by teleporting after the spectate event has been accepted.

Reply:
It's not related, the issue is not about the player being teleported or not, it is always teleported using the spectate command but the camera isn't set. So either you fix this issue or you don't claim to fix it

Reply:
I recommend you to take a look of the video Avoma did that explain the issue on the ticket. Also i would also say that your fix here works only for players (and not mobs since they are recreated) and i think might be broken once player instance are recreated in the future.

Reply:
Possibly, but how am I supposed to resolve the new instance in another world reliably? And more important: How would I detect the world change in the first place without jeopardizing performance too much?

Reply:
I mean scanning every world for an entity UUID is not really efficient...